File: backend/pkg/core/image/service.go
package image

import (
	"context"
	"time"

	"backend/pkg/database/repository"

	"github.com/google/uuid"
)

// Service provides business logic for image generations
type Service struct {
	repo repository.ImageRepository
}

// NewService creates a new image service
func NewService(repo repository.ImageRepository) *Service {
	return &Service{repo: repo}
}

// CreateRequest represents a request to create an image generation
type CreateRequest struct {
	Prompt    string
	Provider  string
	ModelName string
	RequestID string
	MetaData  map[string]interface{}
}

// GetByID retrieves an image generation by its ID
func (s *Service) GetByID(ctx context.Context, id uuid.UUID) (*ImageGeneration, error) {
	return s.repo.FindByID(ctx, id)
}

// GetByRequestID retrieves an image generation by its request ID
func (s *Service) GetByRequestID(ctx context.Context, requestID string) (*ImageGeneration, error) {
	return s.repo.FindByRequestID(ctx, requestID)
}

// Create creates a new image generation record
func (s *Service) Create(ctx context.Context, req CreateRequest) (*ImageGeneration, error) {
	img := &ImageGeneration{
		ID:        uuid.New(),
		Prompt:    req.Prompt,
		Provider:  req.Provider,
		ModelName: req.ModelName,
		RequestID: req.RequestID,
		Status:    StatusPending,
		MetaData:  req.MetaData,
	}

	if err := s.repo.Save(ctx, img); err != nil {
		return nil, err
	}

	return img, nil
}

// MarkCompleted marks an image generation as completed
func (s *Service) MarkCompleted(ctx context.Context, id uuid.UUID, outputURL string, fileIndexID *uuid.UUID) error {
	img, err := s.repo.FindByID(ctx, id)
	if err != nil {
		return err
	}

	now := time.Now()
	img.Status = StatusCompleted
	img.OutputURL = outputURL
	img.FileIndexID = fileIndexID
	img.CompletedAt = &now

	return s.repo.Update(ctx, img)
}

// MarkFailed marks an image generation as failed
func (s *Service) MarkFailed(ctx context.Context, id uuid.UUID, errorMessage string) error {
	img, err := s.repo.FindByID(ctx, id)
	if err != nil {
		return err
	}

	now := time.Now()
	img.Status = StatusFailed
	img.ErrorMessage = errorMessage
	img.CompletedAt = &now

	return s.repo.Update(ctx, img)
}

// GetStatus returns the status of an image generation
func (s *Service) GetStatus(ctx context.Context, id uuid.UUID) (string, error) {
	img, err := s.repo.FindByID(ctx, id)
	if err != nil {
		return "", err
	}
	return img.Status, nil
}

File: backend/src/core/image/service.rs
use std::sync::Arc;

use chrono::Utc;
use uuid::Uuid;

use crate::error::AppError;

use super::{
    CreateImageRequest, IMAGE_STATUS_COMPLETED, IMAGE_STATUS_FAILED, IMAGE_STATUS_PENDING,
    ImageGeneration, ImageRepository,
};

#[derive(Clone)]
pub struct ImageService {
    repository: Arc<dyn ImageRepository>,
}

impl ImageService {
    pub fn new(repository: Arc<dyn ImageRepository>) -> Self {
        Self { repository }
    }

    pub async fn get_by_id(&self, id: Uuid) -> Result<ImageGeneration, AppError> {
        self.repository.find_by_id(id).await
    }

    pub async fn get_by_request_id(&self, request_id: &str) -> Result<ImageGeneration, AppError> {
        self.repository.find_by_request_id(request_id).await
    }

    pub async fn create(&self, request: CreateImageRequest) -> Result<ImageGeneration, AppError> {
        let mut image = ImageGeneration {
            id: Uuid::new_v4(),
            prompt: request.prompt,
            provider: request.provider,
            model_name: request.model_name,
            request_id: request.request_id,
            status: IMAGE_STATUS_PENDING.to_owned(),
            output_url: String::new(),
            file_index_id: None,
            error_message: String::new(),
            meta_data: request.meta_data,
            created_at: None,
            completed_at: None,
        };
        self.repository.save(&mut image).await?;
        Ok(image)
    }

    pub async fn mark_completed(
        &self,
        id: Uuid,
        output_url: String,
        file_index_id: Option<Uuid>,
    ) -> Result<(), AppError> {
        let mut image = self.repository.find_by_id(id).await?;
        image.status = IMAGE_STATUS_COMPLETED.to_owned();
        image.output_url = output_url;
        image.file_index_id = file_index_id;
        image.completed_at = Some(Utc::now());
        self.repository.update(&image).await
    }

    pub async fn mark_failed(&self, id: Uuid, error_message: String) -> Result<(), AppError> {
        let mut image = self.repository.find_by_id(id).await?;
        image.status = IMAGE_STATUS_FAILED.to_owned();
        image.error_message = error_message;
        image.completed_at = Some(Utc::now());
        self.repository.update(&image).await
    }

    pub async fn get_status(&self, id: Uuid) -> Result<String, AppError> {
        self.repository
            .find_by_id(id)
            .await
            .map(|image| image.status)
    }
}
